File: packages/shared-utils/gcw_utils/time.py
"""Time utilities for context signal processing."""
from datetime import datetime, timedelta
from typing import Literal

TimeOfDay = Literal["morning", "lunch", "afternoon", "evening", "night"]
DayType = Literal["weekday", "weekend", "holiday"]


def get_time_of_day(hour: int) -> TimeOfDay:
    """Maps an hour (0-23) to a named time-of-day category."""
    if 6 <= hour < 11:
        return "morning"
    if 11 <= hour < 14:
        return "lunch"
    if 14 <= hour < 18:
        return "afternoon"
    if 18 <= hour < 22:
        return "evening"
    return "night"


def get_day_type(dt: datetime, is_holiday: bool = False) -> DayType:
    """Returns weekday, weekend, or holiday for a given datetime."""
    if is_holiday:
        return "holiday"
    return "weekend" if dt.weekday() >= 5 else "weekday"


def is_within_refresh_window(last_fetch: datetime, now: datetime, window_minutes: int) -> bool:
    """Returns True if last_fetch is still within the refresh window."""
    return (now - last_fetch) < timedelta(minutes=window_minutes)


def format_countdown(seconds_remaining: int) -> str:
    """Formats a countdown in seconds to a human-readable string."""
    if seconds_remaining <= 0:
        return "Expired"
    minutes = seconds_remaining // 60
    seconds = seconds_remaining % 60
    if minutes == 0:
        return f"{seconds}s"
    return f"{minutes}m {seconds}s"

File: packages/shared-utils/src/time.ts
import { TimeOfDay, DayType } from '@gcw/shared-types';

/** Maps an hour (0-23) to a named time-of-day category. */
export function getTimeOfDay(hour: number): TimeOfDay {
  if (hour >= 6 && hour < 11) return 'morning';
  if (hour >= 11 && hour < 14) return 'lunch';
  if (hour >= 14 && hour < 18) return 'afternoon';
  if (hour >= 18 && hour < 22) return 'evening';
  return 'night';
}

/** Returns whether a given Date falls on a weekend. */
export function getDayType(date: Date, isHoliday: boolean): DayType {
  if (isHoliday) return 'holiday';
  const day = date.getDay();
  return day === 0 || day === 6 ? 'weekend' : 'weekday';
}

/** Returns true if two dates are within the same 15-minute window. */
export function isWithinRefreshWindow(
  lastFetch: Date,
  now: Date,
  windowMinutes: number,
): boolean {
  const diffMs = now.getTime() - lastFetch.getTime();
  return diffMs < windowMinutes * 60 * 1000;
}

/** Formats a countdown in seconds to a human-readable string. */
export function formatCountdown(secondsRemaining: number): string {
  if (secondsRemaining <= 0) return 'Expired';
  const minutes = Math.floor(secondsRemaining / 60);
  const seconds = secondsRemaining % 60;
  if (minutes === 0) return `${seconds}s`;
  return `${minutes}m ${seconds}s`;
}
